### PR TITLE
[ENH] Pivot longer improve

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Add more examples to notebook, 
- Fixed spellings and improved sentences
- Add more tests 

**This PR improves #755 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [ ] Check to ensure that test coverage covers the lines of code that you have added.
    - [ ] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [ ] Build the docs locally.
- [ ] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Review by ericmjl — APPROVED** [review]
Wow, this is a wonderful PR, @samukweku! Thank you so much for your work here!

**Comment by ericmjl** [comment]
Let's get a second pair of eyes on the PR. Any takers? If not, I'll merge on Oct 11.

**Comment by samukweku** [comment]
@ericmjl @loganthomas any ideas on how to fix the failing deepsource and codecov? Do I go ahead with the suggestions from DeepSource? A bit hesitant to touch anything, since these were not failing before.

**Comment by ericmjl** [comment]
@samukweku don't worry about them, let's get them fixed in a different PR.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -17,6 +17,8 @@ v0.20.10
   tidyr's ``pivot_longer`` function. Also added an example notebook. @samukweku
 - [ENH] Fixed code to returns error if `fill_value` is not a dictionary. @samukweku
 - [INF] Welcome bot (.github/config.yml) for new users added. Issue #739. @samukweku
+- [ENH] ``pivot_longer`` can handle multiple values in paired columns, and can reshape
+  using a list/tuple of regular expressions in ``names_pattern``. @samukweku
```
Comment (loganthomas) [inline CHANGELOG.rst]: Needs to be updated for `CHANGELOG.rst` that is "on deck"

File: examples/notebooks/Pivoting Data from Wide to Long.ipynb [hunk 2]
```diff
@@ -708,13 +936,176 @@
...
+    ")\n",
+    "\n",
+    "df"
+   ]
+  },
+  {
+   "source": [
+    "The goal is to reshaped the data into long format, with separate columns for `Manufacturer`(Sony,...), `Device`(TV, Radio), `Model`(S3424, ...), maximum quantity and quantity. \n",
+    "\n",
+    "A good solution was proferred, also on [Stack Overflow](https://stackoverflow.com/a/64107688/7175713) :"
```
Comment (loganthomas) [inline examples/notebooks/Pivoting Data from Wide to Long.ipynb]: proffered?

File: janitor/utils.py [hunk 3]
```diff
@@ -665,7 +667,38 @@ def _data_checks_pivot_longer(
...
+                    """
+                    All entries in ``names_pattern`` argument
+                    must be regular expressions.
+                    """
+                )
+            if not isinstance(names_to, (List, Tuple)):
+                raise TypeError(
+                    """
+                ``names_to`` must be a list or tuple.
+                """
```
Comment (loganthomas) [inline janitor/utils.py]: fix indentation to align correctly

File: janitor/utils.py [hunk 4]
```diff
@@ -833,21 +874,54 @@ def _computations_pivot_longer(
...
+
         before_df = _reindex_func(before_df, index)
         between_df = between_df.set_axis(names_to, axis="columns")
 
+        # What should be the appropriate error message
+        # if an inappropriate regular expression is used, that returns
+        # an incorrect number of rows? Where should it be detected?
+        # Personally, I think the onus lies on the user to provide
+        # the right regex.
+
```
Comment (loganthomas) [inline janitor/utils.py]: @samukweku @ericmjl 
Is this still an open topic? If there is a check for the regular expression (i.e. the safeguard check above), I think we are ok here... I agree that the onus should lie on the user to provide the right regex. However, at least some type of check should give the user intuition on why the function could fail. I think this is accomplished with the safeguard check. Just my 2 cents...

**Review by loganthomas — CHANGES_REQUESTED** [review]
@samukweku @ericmjl 
Minimal changes requests (update `CHANGELOG.rst`, one potential typo, one indentation fix, and one comment on a question about raising an error).

I approve, just wanted to capture a few thoughts on potential changes.

Great work here @samukweku ! 😄

File: janitor/utils.py (hunk 4 above)
Comment (samukweku) [reply janitor/utils.py]: I'll take this out. Wanted feedback on that (which I've gotten, thanks)

File: examples/notebooks/Pivoting Data from Wide to Long.ipynb (hunk 2 above)
Comment (samukweku) [reply examples/notebooks/Pivoting Data from Wide to Long.ipynb]: I will change this to something else, which should be clearer.

**Comment by samukweku** [comment]
@loganthomas @ericmjl all fixed now

**Review by loganthomas — APPROVED** [review]

**Pull request merged**